src/main/java/gregtech/client/particle/GTParticleManager.java

```diff
@@ -18,165 +20,178 @@
+ * Singleton class responsible for managing, updating and rendering {@link GTParticle} instances.
  */
 @SideOnly(Side.CLIENT)
 public class GTParticleManager {
-    public final static GTParticleManager INSTANCE = new GTParticleManager();
 
+    public static final GTParticleManager INSTANCE = new GTParticleManager();
+
+    @Nullable
     private static World currentWorld = null;
-    private static final Minecraft mc = Minecraft.getMinecraft();
 
-    private final Map<IGTParticleHandler, ArrayDeque<GTParticle>> renderQueueBack = new HashMap<>();
-    private final Map<IGTParticleHandler, ArrayDeque<GTParticle>> renderQueueFront = new HashMap<>();
-    private final Queue<Tuple<IGTParticleHandler, GTParticle>> newParticleQueue = new ArrayDeque<>();
+    private final Map<IRenderSetup, ArrayDeque<GTParticle>> depthEnabledParticles = new Object2ObjectLinkedOpenHashMap<>();
+    private final Map<IRenderSetup, ArrayDeque<GTParticle>> depthDisabledParticles = new Object2ObjectLinkedOpenHashMap<>();
 
-    public void addEffect(GTParticle... particles) {
-        for (GTParticle particle : particles) {
-            if (particle.getGLHandler() != null) {
-                newParticleQueue.add(new Tuple<>(particle.getGLHandler(), particle));
-            } 
-        }
+    private final List<GTParticle> newParticleQueue = new ArrayList<>();
+
+    public void addEffect(@Nonnull GTParticle particles) {
+        newParticleQueue.add(particles);
     }
 
     public void updateEffects() {
-        updateEffectLayer();
+        if (!depthEnabledParticles.isEmpty()) {
+            updateQueue(depthEnabledParticles);
+        }
+        if (!depthDisabledParticles.isEmpty()) {
+            updateQueue(depthDisabledParticles);
+        }
         if (!newParticleQueue.isEmpty()) {
-            for (Tuple<IGTParticleHandler, GTParticle> handlerParticle = newParticleQueue.poll(); handlerParticle != null; handlerParticle = newParticleQueue.poll()) {
-                IGTParticleHandler handler = handlerParticle.getFirst();
-                GTParticle particle = handlerParticle.getSecond();
-                Map<IGTParticleHandler, ArrayDeque<GTParticle>> renderQueue = particle.getFXLayer() > 0 ? renderQueueFront : renderQueueBack;
-                if (!renderQueue.containsKey(handler)) {
-                    renderQueue.put(handler, new ArrayDeque<>());
-                }
-                ArrayDeque<GTParticle> arrayDeque = renderQueue.get(handler);
-                if (arrayDeque.size() > 6000) {
-                    arrayDeque.removeFirst().setExpired();
+            for (GTParticle particle : newParticleQueue) {
+                var queue = particle.shouldDisableDepth() ? depthDisabledParticles : depthEnabledParticles;
+
+                ArrayDeque<GTParticle> particles = queue.computeIfAbsent(particle.getRenderSetup(), setup -> new ArrayDeque<>());
+
+                if (particles.size() > 6000) {
```

Comment:
Should set this limit to a `public static final` constant.

Alternatively, should this value have a config? Would that functionality be useful to clients at all?

Reply:
Doubt it would be meaningful; 6000 instances of particle in one place is a lot of particles, pretty much unlikely in normal circumstances considering GTParticle's niche and limited usage. Unless a player overamps 6000 blocks of wire at once and then starts assline or something, I think it'll be very hard to see the instance cap actually being a bottleneck.

Since the value is mostly for preventing extreme scenarios like OOM from faulty usage, I don't know if exposing the limit as public const would bring any good, either.